### PR TITLE
[JavaScript] Improve tag write performance

### PR DESCRIPTION
**Pull request by bytemain**
## What do these changes do?

Previously, everytime we write tag, we create new BufferWriter, we alloc new Buffer pool, this is no need.

## Check code requirements

- [x] tests added / passed (if needed)
- [x] Ensure all linting tests pass, see [here](https://github.com/alipay/fury/blob/main/CONTRIBUTING.rst) for how to run them

### CODE REVIEW

File: javascript/packages/fury/lib/classResolver.ts [hunk 1]
```diff
@@ -36,7 +36,8 @@ export default class SerializerResolver {
   };
 
   private readStringPool: string[] = [];
-  private writeStringIndex: string[] = [];
+  private writeStringCount = 0;
+  private writeStringIndex: Record<string, number> = {};
```
Comment (wangweipeng2) [inline javascript/packages/fury/lib/classResolver.ts:40]: I have a idea, Perhaps, we can generate tag accessors code at runtime, 
```
```


I have an idea. Perhaps we can generate tag accessor code at runtime. Here's how we might do it:


1. Add a counter to classResolver:
```
writeStringCount = 0;
```

2. Generate a closure when registering a tag:

```

function createTagWriter() {
    retindex = -1;

    return function () {
        // store the index
        if (index > -1) {
            writeTagIndex(); // USESTRINGID

        } else {
            writeTagContent(); // USESTRINGVALUE
            index = writeStringCount++;
        }
    } 
}

```

3. When writing the tag, we can directly call the result of the createTagWriter function, which can help us avoid the hashmap lookup.

File: javascript/packages/fury/lib/classResolver.ts [hunk 2]
```diff
@@ -74,7 +75,7 @@ export default class SerializerResolver {
 
   reset() {
     this.readStringPool = [];
-    this.writeStringIndex = [];
+    this.writeStringIndex = {};
```
Comment (wangweipeng2) [inline javascript/packages/fury/lib/classResolver.ts:78]: Should clear the writeStringCount

File: javascript/packages/fury/lib/classResolver.ts [hunk 3]
```diff
@@ -94,30 +94,30 @@ export default class SerializerResolver {
...
   }
 
-  tagToBuffer(tag: string) {
-    const tagBuffer = fromUint8Array(new TextEncoder().encode(tag));
-    let tagHash = x64hash128(tagBuffer, 47).getBigUint64(0);
-    if (tagHash === BigInt(0)) {
-      tagHash = BigInt(1);
-    }
+  createTagWriter(tag: string) {
+    let tagIndex = -1;
```
Comment (wangweipeng2) [inline javascript/packages/fury/lib/classResolver.ts:98]: Each tag call createTagWriter one time, so the tagIndex should be reset before start serialize. Perhaps we can store 
 the tagIndex on classResolver, and reset it to zero when start serialize.
The code maybe like this:
```javascript
 this.writeStringIndex = [];
 reset() {
    this.writeStringIndex.fill(-1);
 }
 createTagWriter(tag: string) {
    this.writeStringIndex.push(-1);
    let idx = this.writeStringIndex.length - 1;
    return () =>  {
        const tagIndex = this.writeStringIndex[idx];
       // when update
       this.writeStringIndex[idx] = this.writeStringCount++;
    }
}
```

Comment (bytemain) [reply javascript/packages/fury/lib/classResolver.ts]: done

**Review by theweipeng — APPROVED** [review]
Great work

**Pull request merged**